earth_enterprise/src/scons/getversion.py

```diff
@@ -42,16 +43,125 @@ def GetLongVersion(backupFile, label=''):
 
 
 def _GitGeneratedLongVersion():
-    """Take the raw information parsed by git, and use it to
-       generate an appropriate version string for GEE."""
+    """Calculate the version name and build number into a single build string."""
 
-    raw = _GetCommitRawDescription()
+    versionName, buildNumber = _GitVersionNameAndBuildNumber()
+    return "{0}-{1}".format(versionName, buildNumber)
 
-    # For tagged commits, return the tag itself
-    if _IsCurrentCommitTagged(raw):
-        return _VersionForTaggedHead(raw)
+
+def _GitCommitCount(tagName='HEAD', baseRef=''):
+    """calculate git commit counts"""
+    repo = _GetRepository()
+    if not baseRef:
+        return len(list(repo.iter_commits(tagName)))
+    else:
+        return len(list(repo.iter_commits(baseRef + '..' + tagName)))
+
+
+def _GitVersionNameAndBuildNumber():
+    """Get the version name and build number based on state of git
+    Use a tag only if HEAD is directly pointing to it and it is a
+    release build tag (see _GetCommitRawDescription for details)
+    otherwise use the branch name"""
+
+    # For tagged commits use the tag
+    rawDescription = _GetCommitRawDescription()
+    if _IsCurrentCommitTagged(rawDescription):
+        # Extract version name and build number
+        # from the tag (should be a release build tag)
+        splitTag = rawDescription.split('-')
+        return splitTag[0], splitTag[1]
     else:
-        return _VersionFromTagHistory(raw)
+        # Use branch name if we are not a detached HEAD
+        branchName = _GitBranchName()
+        if not branchName:
+            # we are a detached head not on a tag so just treat the
+            # raw describe like a topic branch name
+            return rawDescription, str(_GitCommitCount())
+        else:
+            # Get the version name from the branch name
+            if _IsReleaseBranch(branchName):
+                versionName = _GetReleaseVersionName(branchName)
+                return versionName, '{0}.{1}'.format(_GitBranchedBuildNumber(versionName), _GitCommitCount('HEAD', versionName))
+            else:
+                return _sanitizeBranchName(branchName),  str(_GitCommitCount())
+
+
+def _GitBranchedBuildNumber(versionName):
+    """Returns what the build number was from the branch point"""
+    prevRelTag = _GitPreviousReleaseTag(versionName)
+    prevBuildNumber = ''
+
+    if prevRelTag:
+        prevBuildNumber = prevRelTag.split('-')[1]
+    else:
+        prevBuildNumber = str(_GitCommitCount(versionName))
+
+    return prevBuildNumber
+
+
+def _GitPreviousReleaseTag(versionName):
+    """Looks for the tail tag for this versionName (the tag with the same name as the version name)
+    and if it finds it then it looks for any release build tags that are also pointing to the same
+    commit"""
+    tags = _GetRepository().tags
+    tailTag = next((tag for tag in _GetRepository().tags if tag.name == versionName), None)
```

Comment:
The logic in this line is a little confusing. I'd suggest splitting it into multiple lines

Reply:
The new code is much easier to understand